### DISCUSSION TITLE
Collapse / Toggle Block Type

### DISCUSSION THREAD

**Discussion by gotjoshua** [toeverything/blocksuite#906] [Ideas]
- [ ] Agree on relevance
- [ ] Agree on style
- [ ] Agree that it belongs here in [packages/blocks/src](https://github.com/toeverything/blocksuite/tree/master/packages/blocks/src)
- [ ] Agree that it is an appropriate task for a first time contributor 

| Example from Workflowy        | Example from Notion           | 
| ------------- |:-------------:| 
| ![telegram-cloud-photo-size-4-5960671904379551343-y](https://user-images.githubusercontent.com/8686454/215497247-4e0ac5aa-f98c-4143-aa9a-a2f90fcec179.jpg)    |![telegram-cloud-photo-size-4-5960671904379551396-x](https://user-images.githubusercontent.com/8686454/215497320-e8aa87ee-03de-4ed3-99ca-16cfe33e7bad.jpg) |

**Comment by doodlewind** [comment]
Hi, thanks for your feedback!

I think this is a very interesting feature, and it's also something we are going to partially covered! In our design of the DatabaseBlock (see #629), we will be able to render nest-able blocks (e.g., nested list) as sub-row in table. Folding and unfolding the row could naturally be supported when evolving the DatabaseBlock (and ListBlock).

But to my knowledge, it may be a bit hard for a first -time contributor to get things right in this feature for now, since there will also be some non-trivial modification required in the containers (e.g., page block and database block) UI and selection (deciding which block should be picked, we handle this in a more "manual" way in `selection.ts`).

So before developing this feature, I think this could be better to collect more feedback in discussion, so I'm going to convert it. Hope to see more thoughts about this!

**Reply by gotjoshua** [reply]
> a bit hard for a first -time contributor

i'm super happy if the team takes it on!
I think a pure collapse/toggle component is needed and will need to be able to perform within a normal text context and within a database row rendering context..

agreed?

**Comment by gotjoshua** [comment]
Couldn't it be an additional type for the list block?

https://github.com/toeverything/blocksuite/blob/ff10b69c698a5cce4c434ffaeb8b5fca51744d8a/packages/blocks/src/list-block/utils/get-list-icon.ts#L34

**Reply by HeJiachen-PM** [reply]
I think a 'collapse' feature is not bad for titles. However I am not sure whether we should do a 'toggle list' or make normal unordered list collapseable

**Reply by gotjoshua** [reply]
Well i think any nested list can be a mix of elements that play nice together

Eg top level collapse with task list(s) as  children

**Reply by HeJiachen-PM** [reply]
> Well i think any nested list can be a mix of elements that play nice together
> 
> Eg top level collapse with task list(s) as  children 

True

**Reply by doodlewind** [reply]
Yep things will be easier if we use a standalone list type, instead of making the current list types foldable.

**Reply by gotjoshua** [reply]
i have a starting point for that on my fork... 

https://github.com/toeverything/blocksuite/pull/908

i know you closed it as not planned... but i want to play around anyway. No expectations.
Let it serve as a discussion piece.

**Reply by gotjoshua** [reply]
[Vercel preview](https://blocksuite-react-git-fork-onezoomin-feat-to-abf8bc-toeverything.vercel.app/)

typing '>>' [space] will create the collapse element

i still need to toggle open/closed correctly on click (for now it only choses the correct icon if the node already has a child when you type >>
![Screen Shot 2023-01-30 at 6 41 23 PM](https://user-images.githubusercontent.com/8686454/215566765-20d5aaef-52ec-4729-8750-6931fb8415cf.png)

**Reply by gotjoshua** [reply]
![ToggleDemo](https://user-images.githubusercontent.com/8686454/215579747-e6e5b884-eb86-4cca-bef4-bb0da8f721e9.gif)

**Comment by gotjoshua** [comment]
Here is another UI example from Logseq - similar to workflowy, except that the arrow is only shown on hover (in this example on the 3rd line "Collapsed with kids"

![LogseqExample](https://user-images.githubusercontent.com/8686454/216107164-f3a34484-8bd3-453b-bfb9-3fbbd9e69673.png)

note: the pathway to the currently focussed block is a plugin.
![FurtherPathway](https://user-images.githubusercontent.com/8686454/216108034-1d8baa08-5735-42de-ba44-42afab4d5869.png)

**Comment by houjoe0829** [comment]
We shared a discussion about Toggle design in the community, please take a look and feel free to leave your comments.
https://community.affine.pro/c/build-in-public/what-is-affine-s-toggle-block
